### PR TITLE
Added default color dict check when automatically selecting the color mode

### LINKED ISSUE #2953: Creating a labels layer from the layer `data_tuple` results in empty colors.
## 🐛 Bug

Copying labels through the `data_tuple` API results in a new Labels layers with empty colors.
I'm not sure if this a real bug because I'm using some "private" methods.

## To Reproduce

```python3
import napari
import numpy as np

viewer = napari.Viewer()
labels = viewer.add_labels(np.random.randint(10, size=(10, 10)))

data_tuple = labels.as_layer_data_tuple()
cpy = viewer._add_layer_from_data(*data_tuple)[0]

# cpy.color_mode = labels.color_mode  # with this line it works                                                                                                                                                                                          

napari.run()
```

## Expected behavior

The new layer should be equal to its source, including the color mapping.

## Environment

```
napari: 0.4.8
Platform: Linux-5.8.0-55-generic-x86_64-with-glibc2.10
System: Ubuntu 20.04.2 LTS
Python: 3.8.8 (default, Apr 13 2021, 19:58:26) [GCC 7.3.0]
Qt: 5.15.2
PyQt5: 5.15.4
NumPy: 1.20.3
SciPy: 1.6.3
Dask: 2021.04.0
VisPy: 0.6.6

OpenGL:
- GL version: 4.6 (Compatibility Profile) Mesa 20.2.6
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 3440x1440, scale 1.0

Plugins:
- console: 0.0.3
- napari-dexp: 0.1.dev7+g3714459
- napari-simple-crop: 0.1.dev1+gd89c7ea.d20210629
- ome_zarr: 0.0.19
- scikit-image: 0.4.8
- svg: 0.1.4
```

## Additional context

I thought this could be solved by adding this attribute to the [_get_state](https://github.com/napari/napari/blob/d310847f0c79f062daadecb8a159e8972c074470/napari/layers/labels/labels.py#L514) but it does not work with the `_add_layer_from_data` since it is not an argument of the `add_labels` method.

### PR DESCRIPTION
**Pull request by JoOkuma**
# Description
The `Labels` layer color setter selected the `color_mode` according to the `color` dict content, but it did not take into account that it could contain only the default values, resulting in undesired behaviors when using the layers' `data_tuple` API.

## Type of change
- [X] Bug-fix (non-breaking change which fixes an issue)

# References
Closes #2953.

# How has this been tested?
- [X] the example from the issue results in the expected behavior.
- [X] passes the labels layer test suite.

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by Carreau** [comment]
Rebased and fixed conflicts (it was just another added test in the same place)

**Review by Carreau — APPROVED** [review]
Not an expert in this area, but +1.

**Comment by sofroniewn** [comment]
Would this fix #2479 too @JoOkuma, seems like it might be the same issue. Maybe @alisterburt can review too - if he likes it this change is fine by me, though maybe  `_is_default_color` can get a full doc string too.

I will say I find this part of the code quite hard to reason about - I wonder if we wouldn't be better off pulling out the `color_mode` change into its own setter, this might mean some functionality now needs two lines of code to enable instead of one, but it might make the logic and flow simpler. Thoughts @jni?

File: napari/layers/labels/_tests/test_labels.py [hunk 1]
```diff
@@ -978,3 +978,9 @@ def test_paint_3d_negative_scale(scale):
     np.testing.assert_array_equal(
         np.sum(labels_layer.data, axis=(1, 2, 3)), [0, 95, 0]
     )
+
+
+def test_is_default_color():
```
Comment (alisterburt) [inline napari/layers/labels/_tests/test_labels.py:983]:
```suggestion
def test_is_default_color():
    """Test that Labels._is_default_color() returns True for the default color 
    value of the Labels layer.
    """
```

File: napari/layers/labels/labels.py [hunk 2]
```diff
@@ -416,7 +416,7 @@ def color(self):
     @color.setter
     def color(self, color):
 
-        if not color:
+        if not color or self._is_default_color(color):
```
Comment (alisterburt) [inline napari/layers/labels/labels.py:424]:
```suggestion
        if not color or self._is_default_color(color):  # default check fixes #2479 and #2953 
```

File: napari/layers/labels/labels.py [hunk 3]
```diff
@@ -436,6 +436,24 @@ def color(self, color):
         self._color = colors
         self.color_mode = color_mode
 
+    def _is_default_color(
+        self, color: Dict[Optional[int], np.ndarray]
+    ) -> bool:
```
Comment (alisterburt) [inline napari/layers/labels/labels.py:444]:
```suggestion
    def _is_default_color(
        self, color: Dict[Optional[int], np.ndarray]
    ) -> bool:
        """Checks whether a color is equal to the default color value of the layer."""
```

**Review by alisterburt — APPROVED** [review]
Great job @JoOkuma ! - I checked and this closes #2479 too

I've added a couple of suggestions for docstrings, feel free to modify as you see fit!

I agree with @sofroniewn that it's a little hard to follow so I also added a comment where the default check is made in the `color` setter pointing to these issues, again feel free to modify

**Comment by JoOkuma** [comment]
@alisterburt thanks for the input

**Comment by JoOkuma** [comment]
> Would this fix #2479 too @JoOkuma, seems like it might be the same issue. Maybe @alisterburt can review too - if he likes it this change is fine by me, though maybe `_is_default_color` can get a full doc string too.
> 
> I will say I find this part of the code quite hard to reason about - I wonder if we wouldn't be better off pulling out the `color_mode` change into its own setter, this might mean some functionality now needs two lines of code to enable instead of one, but it might make the logic and flow simpler. Thoughts @jni?

I'm not completely sure what would be the benefits of having its own setter, because I think we would still need to check if it still has the default values or not.

An alternative would be to change the getter of `color` to add the default values only when it is called, so `_color` would not be changed through some side effects, but I'm not sure if this could lead to some other issues.

File: napari/layers/labels/labels.py [hunk 4]
```diff
@@ -436,6 +437,25 @@ def color(self, color):
...
+        """Checks whether a color is equal to the default color value of the layer."""
+        if len(color) != 2:
+            return False
+
+        default_color = {
+            None: transform_color('black')[0],
+            self._background_label: transform_color('transparent')[0],
+        }
```
Comment (sofroniewn) [inline napari/layers/labels/labels.py]: this should get defined once and be a private attribute on the layer

**Review by sofroniewn — COMMENTED** [review]
> I'm not completely sure what would be the benefits of having its own setter, because I think we would still need to check if it still has the default values or not.

I guess it is because it would avoid the side effect of changing one property causing a change in another. But i'm also not sure we'd still have a problem. The issue is that setting the color explicitly changes color mode from `auto` to `direct`, this would no longer be the case, you'd have to make two calls. No? I havn't used this code in a while so I might be missing something, but that seems preferable to me

> An alternative would be to change the getter of color to add the default values only when it is called, so _color would not be changed through some side effects, but I'm not sure if this could lead to some other issues.

I think i'd rather avoid complex logic in the getter too.

**Comment by sofroniewn** [comment]
Something kind of like

```python
    @property
    def color(self):
        """dict: custom color dict for label coloring"""
        return self._color

    @color.setter
    def color(self, color):

        # default check fixes #2479 and #2953
        if not color:
            color = {}

        if self._background_label not in color:
            color[self._background_label] = 'transparent'

        if None not in color:
            color[None] = 'black'

        colors = {
            label: transform_color(color_str)[0]
            for label, color_str in color.items()
        }

        self._color = colors

    @property
    def color_mode(self):
        """......."""
        return str(self._color_mode)

    @color_mode.setter
    def color_mode(self, color_mode):
        self.color_mode = LabelColorMode(color_mode)
        self.events.color_mode()
```

File: napari/layers/labels/labels.py (hunk 4 above)
Comment (JoOkuma) [reply napari/layers/labels/labels.py]: Done, thanks for the suggestions

**Comment by JoOkuma** [comment]
Now I understood your suggestion for the `color_mode` setter.

From the user perspective, I kind of prefer picking the color_mode automatically and for now, it is not very complicated.

But if you guys prefer I can implement the `color_mode` setter as suggested by Nick.

**Comment by sofroniewn** [comment]
> But if you guys prefer I can implement the color_mode setter as suggested by Nick.

@jni can you weigh in on this one, thanks!!
